pkg/controller/argocd/role.go

```diff
@@ -0,0 +1,177 @@
+	sa.Name = name
+
+	lbls := sa.ObjectMeta.Labels
+	lbls[common.ArgoCDKeyName] = name
+	sa.ObjectMeta.Labels = lbls
+
+	return sa
+}
+
+// newRole returns a new ServiceAccount instance.
+func newRole(name string, cr *argoprojv1a1.ArgoCD) *v1.Role {
+	return &v1.Role{
+
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      name,
+			Namespace: cr.Namespace,
+			Labels:    labelsForCluster(cr),
+		},
+	}
+}
+
+// newClusterRoleWithName creates a new ClusterRole with the given name for the given ArgCD.
+func newClusterRoleWithName(name string, cr *argoprojv1a1.ArgoCD) *v1.ClusterRole {
+	sa := newClusterRole(name, cr)
+	sa.Name = generateResourceName(name, cr)
+
+	lbls := sa.ObjectMeta.Labels
+	lbls[common.ArgoCDKeyName] = name
+	sa.ObjectMeta.Labels = lbls
+
+	return sa
+}
+
+// newClusterRole returns a new ClusterRole instance.
+func newClusterRole(name string, cr *argoprojv1a1.ArgoCD) *v1.ClusterRole {
+	return &v1.ClusterRole{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:   generateResourceName(name, cr),
+			Labels: map[string]string{},
+		},
+	}
+}
+
+func allowedNamespace(current string, configuredList string) bool {
+	isAllowedNamespace := false
+	if configuredList != "" {
+		if configuredList == "*" {
+			isAllowedNamespace = true
+		} else {
+			namespaceList := strings.Split(configuredList, ",")
+			for _, n := range namespaceList {
+				if n == current {
+					isAllowedNamespace = true
+				}
+			}
+		}
+	}
+	return isAllowedNamespace
+}
+
+// reconcileRoles will ensure that all ArgoCD Service Accounts are configured.
+func (r *ReconcileArgoCD) reconcileRoles(cr *argoprojv1a1.ArgoCD) error {
+	if err := r.reconcileRole(applicationController, policyRuleForApplicationController(), cr); err != nil {
+		return err
+	}
+
+	if err := r.reconcileRole(dexServer, policyRuleForDexServer(), cr); err != nil {
+		return err
+	}
+
+	if err := r.reconcileRole(server, policyRuleForServer(), cr); err != nil {
+		return err
+	}
+
+	if err := r.reconcileRole(redisHa, policyRuleForRedisHa(), cr); err != nil {
+		return err
+	}
+
+	rules := []v1.PolicyRule{}
+
+	if cr.Spec.ManagementScope.Cluster != nil && *cr.Spec.ManagementScope.Cluster {
+		if os.Getenv("ARGOCD_CLUSTER_CONFIG_ENABLED") == "true" &&
+			allowedNamespace(cr.Namespace, os.Getenv("ARGOCD_CLUSTER_CONFIG_NAMESPACES")) {
+			rules = append(rules, policyRoleForClusterConfig()...)
+		}
+	}
+
+	rules = append(rules, policyRuleForApplicationControllerClusterRole()...)
+	if err := r.reconcileClusterRole(applicationController, rules, cr); err != nil {
+		return err
+	}
+
+	if err := r.reconcileClusterRole(server, policyRuleForServerClusterRole(), cr); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// reconcileClusterRole
+func (r *ReconcileArgoCD) reconcileClusterRole(name string, policyRules []v1.PolicyRule, cr *argoprojv1a1.ArgoCD) error {
+	rbacClient := r.kc.RbacV1()
+
+	role, err := rbacClient.ClusterRoles().Get(context.TODO(), generateResourceName(name, cr), metav1.GetOptions{})
```

Comment:
we need to replace generateResourceName with `name` as generateResourceName is only used to generate local role name not cluster role name.

// code that is working for me
role, err := rbacClient.ClusterRoles().Get(context.TODO(), name, metav1.GetOptions{})